File: iosApp/iosApp/Domain/Generator/EaspbvQrBuilder.swift
import Foundation

enum QrGenerationError: LocalizedError {
    case missingKeyOrMerchantCode
    case invalidName
    case invalidCity
    case invalidAmount

    var errorDescription: String? {
        switch self {
        case .missingKeyOrMerchantCode: return "Ingresa una llave o un código de comercio"
        case .invalidName: return "Nombre obligatorio, máximo 25 caracteres"
        case .invalidCity: return "Ciudad obligatoria, máximo 15 caracteres"
        case .invalidAmount: return "Monto inválido (solo dígitos y hasta 2 decimales)"
        }
    }
}

/// Construye el rawText de un QR estático según el estándar EASPBV v1.4-2025.
/// Espejo del builder de la app Android: tags 00, 01, 26, 49, 50, 52, 53,
/// 54 (opcional), 58, 59, 60, 62 y 63 (CRC).
enum EaspbvQrBuilder {

    private static let terminalLabel = "TERMQR01"
    private static let purposePurchase = "00"

    static func build(_ params: QrGenerationParams) throws -> String {
        try validate(params)
        let tlv = TlvEncoder.tlv

        var payload = ""
        payload += tlv("00", "01")
        payload += tlv("01", "11")

        if let keyType = params.keyType,
           let keyValue = params.keyValue?.trimmingCharacters(in: .whitespaces),
           !keyValue.isEmpty {
            payload += tlv(
                "26",
                tlv("00", "\(params.network.guidPrefix).LLA") + tlv(keyType.subTag, keyValue)
            )
        }

        payload += tlv(
            "49",
            tlv("00", "\(params.network.guidPrefix).RED") + tlv("01", params.network.networkId)
        )

        if let merchantCode = params.merchantCode?.trimmingCharacters(in: .whitespaces),
           !merchantCode.isEmpty {
            payload += tlv(
                "50",
                tlv("00", "\(params.network.guidPrefix).CU") + tlv("01", merchantCode)
            )
        }

        payload += tlv("52", params.mcc)
        payload += tlv("53", "170")
        if let amount = params.amount?.trimmingCharacters(in: .whitespaces), !amount.isEmpty {
            payload += tlv("54", amount)
        }
        payload += tlv("58", "CO")
        payload += tlv("59", params.merchantName.trimmingCharacters(in: .whitespaces))
        payload += tlv("60", params.merchantCity.trimmingCharacters(in: .whitespaces))
        payload += tlv("62", tlv("07", terminalLabel) + tlv("08", purposePurchase))

        let toChecksum = payload + "6304"
        return toChecksum + Crc16CcittFalse.checksum(toChecksum)
    }

    private static func validate(_ params: QrGenerationParams) throws {
        let hasKey = params.keyType != nil
            && !(params.keyValue?.trimmingCharacters(in: .whitespaces).isEmpty ?? true)
        let hasMerchantCode = !(params.merchantCode?.trimmingCharacters(in: .whitespaces).isEmpty ?? true)
        guard hasKey || hasMerchantCode else { throw QrGenerationError.missingKeyOrMerchantCode }

        let name = params.merchantName.trimmingCharacters(in: .whitespaces)
        guard !name.isEmpty, name.count <= 25 else { throw QrGenerationError.invalidName }

        let city = params.merchantCity.trimmingCharacters(in: .whitespaces)
        guard !city.isEmpty, city.count <= 15 else { throw QrGenerationError.invalidCity }

        if let amount = params.amount?.trimmingCharacters(in: .whitespaces), !amount.isEmpty {
            let pattern = "^\\d{1,10}(\\.\\d{1,2})?$"
            guard amount.range(of: pattern, options: .regularExpression) != nil else {
                throw QrGenerationError.invalidAmount
            }
        }
    }
}

File: androidApp/src/main/kotlin/dev/code93/emvqr/domain/generator/EaspbvQrBuilder.kt
package dev.code93.emvqr.domain.generator

import dev.code93.emvqr.domain.generator.TlvEncoder.tlv
import dev.code93.emvqr.domain.model.QrGenerationParams

/**
 * Construye el rawText de un QR estático según el estándar EASPBV v1.4-2025.
 *
 * Orden de tags: 00, 01, 26 (llave), 49 (red), 50 (código comercio), 52, 53,
 * 54 (opcional), 58, 59, 60, 62 (terminal + propósito) y 63 (CRC).
 */
object EaspbvQrBuilder {

    private const val TERMINAL_LABEL = "TERMQR01"
    private const val PURPOSE_PURCHASE = "00"

    fun build(params: QrGenerationParams): String {
        validate(params)

        val payload = buildString {
            append(tlv("00", "01"))                       // Payload Format Indicator
            append(tlv("01", "11"))                       // QR estático (con monto = híbrido)

            if (params.keyType != null && !params.keyValue.isNullOrBlank()) {
                append(
                    tlv(
                        "26",
                        tlv("00", "${params.network.guidPrefix}.LLA") +
                            tlv(params.keyType.subTag, params.keyValue.trim())
                    )
                )
            }

            append(
                tlv(
                    "49",
                    tlv("00", "${params.network.guidPrefix}.RED") +
                        tlv("01", params.network.networkId)
                )
            )

            if (!params.merchantCode.isNullOrBlank()) {
                append(
                    tlv(
                        "50",
                        tlv("00", "${params.network.guidPrefix}.CU") +
                            tlv("01", params.merchantCode.trim())
                    )
                )
            }

            append(tlv("52", params.mcc))
            append(tlv("53", "170"))                      // COP, ISO 4217
            params.amount?.takeIf { it.isNotBlank() }?.let { append(tlv("54", it.trim())) }
            append(tlv("58", "CO"))
            append(tlv("59", params.merchantName.trim()))
            append(tlv("60", params.merchantCity.trim()))
            append(tlv("62", tlv("07", TERMINAL_LABEL) + tlv("08", PURPOSE_PURCHASE)))
        }

        val toChecksum = payload + "6304"
        return toChecksum + Crc16CcittFalse.checksum(toChecksum)
    }

    private fun validate(params: QrGenerationParams) {
        val hasKey = params.keyType != null && !params.keyValue.isNullOrBlank()
        val hasMerchantCode = !params.merchantCode.isNullOrBlank()
        require(hasKey || hasMerchantCode) { "Se requiere una llave o un código de comercio" }
        require(params.merchantName.isNotBlank()) { "El nombre del comercio es obligatorio" }
        require(params.merchantName.trim().length <= 25) { "Nombre máximo 25 caracteres" }
        require(params.merchantCity.isNotBlank()) { "La ciudad es obligatoria" }
        require(params.merchantCity.trim().length <= 15) { "Ciudad máximo 15 caracteres" }
        require(params.mcc.matches(Regex("\\d{4}"))) { "El MCC debe tener 4 dígitos" }
        params.amount?.takeIf { it.isNotBlank() }?.let {
            require(it.trim().matches(Regex("\\d{1,10}(\\.\\d{1,2})?"))) { "Monto inválido" }
        }
    }
}
